readthedocs/projects/tests/mockers.py

```diff
@@ -0,0 +1,255 @@
+import shutil
+import os
+
+from unittest import mock
+
+from django.conf import settings
+
+from readthedocs.builds.constants import BUILD_STATE_TRIGGERED
+from readthedocs.projects.constants import MKDOCS
+
+
+class BuildEnvironmentMocker:
```

Comment:
This class is the most complex thing from this PR, unfortunately. It's _really_ hard to test our code because it requires a remote server running an API _and_ Docker to execute the commands.

This code is the most accurate attempt that I was able to do to mock these calls in a granular way making sure that we can execute as much code as possible. However, it's not perfect and there are inner functions that I wasn't able to mock properly yet. For example, I wrote this problem for this case in https://github.com/readthedocs/readthedocs.org/blob/d5a8ffb0e458e447772bc7d6008e5a39893f3af7/readthedocs/projects/tests/mockers.py#L72-L79

This class is similar to the work tried in `EnvironmentMockGroup` at https://github.com/readthedocs/readthedocs.org/blob/d5a8ffb0e458e447772bc7d6008e5a39893f3af7/readthedocs/rtd_tests/mocks/environment.py#L7 but covers all the build's process. The new class should replace the old one.

I'm open to receiving feedback here and finding out a better way to do this. I had a really hard time dealing with mocks and trying to make the full build process "runable" from a test.

Reply:
We went over this in the meeting and all my questions so far were covered :+1: 

I liked the approach of testing the model updates through the API. This is more helpful that testing model state at the begin/end of a test. I also appreciate how it is closer to an integration test in that we're not testing each individual command executions, but *the outcome* from executing commands and flow through our build process.

As with any mocking, it's hard to track what is not mocked, but we (probably more like developers in general) have yet to come up with a good answer to that mystery.